third_party/move/move-vm/types/src/module_id_interner.rs

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::interner::BTreeInterner;
+use move_core_types::language_storage::ModuleId;
+use parking_lot::RwLock;
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
+pub struct InternedModuleId(pub usize);
+
+pub struct InternedModuleIdPool {
+    inner: RwLock<BTreeInterner<ModuleId>>,
+}
+
+impl InternedModuleIdPool {
+    pub fn new() -> Self {
+        Self {
+            inner: RwLock::new(BTreeInterner::new()),
+        }
+    }
+
+    pub fn intern(&self, module_id: ModuleId) -> InternedModuleId {
```

Comment:
nit: would passing by reference API also be useful? So we avoid clone if already interned

Reply:
Let me see. Conceptually we want something like this:
https://docs.rs/hashbrown/latest/hashbrown/struct.HashMap.html#method.raw_entry_mut